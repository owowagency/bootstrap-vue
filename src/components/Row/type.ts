import {Prop, PropType} from 'vue';

const aligns = ['baseline', 'center', 'end', 'start', 'stretch'] as const;

type Align = typeof aligns[number];

const alignProp: Prop<Align> = {
    type: String as PropType<Align>,
    default: undefined,
    validator: (a: Align) => aligns.includes(a),
};

const gutterProp: Prop<number | string> = {
    type: [Number, String],
    default: undefined,
};

const justifies = ['around', 'between', 'center', 'end', 'evenly', 'start'] as const;

type Justify = typeof justifies[number];

const justifyProp: Prop<Justify> = {
    type: String as PropType<Justify>,
    default: undefined,
    validator: (j: Justify) => justifies.includes(j),
};

export {
    Align,
    alignProp,
    aligns,
    gutterProp,
    Justify,
    justifies,
    justifyProp,
};
