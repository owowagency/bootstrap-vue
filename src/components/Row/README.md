# Row

```vue
<Row align-sm="center" align-xl="start" row-cols>
    <Col>...</Col>

    <Col>...</Col>
</Row>
```

## Imported Props

| Prop name      | Description                                                                       | Type            | Values | Default   |
| -------------- | --------------------------------------------------------------------------------- | --------------- | ------ | --------- |
| align[Size]    | The alignment of the children. Results in `.align-items-[size]-*`                 | `string`        | -      | undefined |
| cols[Size]     | The amount of cols the column should span. Results in `.row-cols-[size]-*`        | `int`\|`string` | -      | undefined |
| gutters[Size]  | The amount of cols the column should span. Results in `.g-[size]-*`               | `int`\|`string` | -      | undefined |
| guttersX[Size] | The amount of cols the column should span. Results in `.g-x-[size]-*`             | `int`\|`string` | -      | undefined |
| guttersY[Size] | The amount of cols the column should span. Results in `.g-y-[size]-*`             | `int`\|`string` | -      | undefined |
| justify[Size]  | The amount of cols the column should span. Results in `.justify-content-[size]-*` | `string`        | -      | undefined |
| tag            | The html tag that will be used for the component                                  | `string`        | -      | 'div'     |

## Props

| Prop name | Description | Type        | Values | Default |
| --------- | ----------- | ----------- | ------ | ------- |
| align     |             | alignProp   | -      |         |
| cols      |             | colProp     | -      |         |
| gutters   |             | gutterProp  | -      |         |
| guttersX  |             | gutterProp  | -      |         |
| guttersY  |             | gutterProp  | -      |         |
| justify   |             | justifyProp | -      |         |

## Slots

| Name    | Description                     | Bindings |
| ------- | ------------------------------- | -------- |
| default | Displays content inside the row |          |
