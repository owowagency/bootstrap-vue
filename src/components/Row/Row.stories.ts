import Row, {aligns, justifies} from '.';
import Col from '@/components/Col';
import Container from '@/components/Container';
import {breakpoints} from '@/composables/useBreakpoints';

const numberArg = {control: {type: 'number'}};

export default {
    title: 'Layout/Row',
    component: Row,
    argTypes: {
        cols: numberArg,
        ...Object.fromEntries(breakpoints.map(b => [`cols-${b}`, numberArg])),
        gutters: numberArg,
        ...Object.fromEntries(breakpoints.map(b => [`gutters-${b}`, numberArg])),
        guttersX: numberArg,
        ...Object.fromEntries(breakpoints.map(b => [`gutters-x-${b}`, numberArg])),
        guttersY: numberArg,
        ...Object.fromEntries(breakpoints.map(b => [`gutters-y-${b}`, numberArg])),
    },
};

const template = (args) => ({
    components: {Col, Container, Row},
    setup: () => ({args}),
    template: `
        <Container class="p-3 border">
            <Row v-bind="args">
                <Col
                    v-for="i in 2"
                    class="border bg-light"
                >
                    Column {{ i }}
                </Col>
            </Row>
        </Container>
    `,
});

export const Cols = template.bind({});

Cols.args = {cols: 6};

export const Justify = () => ({
    components: {Col, Container, Row},
    setup: () => ({justifies}),
    template: `
        <Container>
            <Row
                v-for="justify in justifies"
                class="mb-3 p-3 border"
                :justify="justify"
            >
                <Col
                    v-for="i in 2"
                    cols="4"
                    class="border bg-light"
                >
                    {{ justify }}
                </Col>
            </Row>
        </Container>
    `,
});

export const Align = () => ({
    components: {Col, Container, Row},
    setup: () => ({aligns}),
    template: `
        <Container>
            <Row
                v-for="align in aligns"
                :align="align"
                class="mb-3 p-3 border"
                style="height: 10rem"
            >
            <Col
                v-for="i in 3"
                cols="4"
                class="border bg-light"
            >
                {{ align }}
            </Col>
            </Row>
        </Container>
    `,
});
