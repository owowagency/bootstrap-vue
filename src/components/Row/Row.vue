<template>
    <component
        :is="tag"
        class="row"
        :class="classes"
    >
        <slot />
    </component>
</template>

<script lang="ts">
import {alignProp, gutterProp, justifyProp} from '@/components/Row/type';
import {computed, toRef} from 'vue';
import useBreakpoints, {Breakpoint, breakpointProps} from '@/composables/useBreakpoints';
import {colProp} from '@/composables/useCol';
import {tagProps} from '@/composables/useTag';
import useClasses from '@/composables/useClasses';
import useStringTemplate from '@/composables/useStringTemplate';

const propsAlign = breakpointProps(alignProp, 'align-{0}') as Record<`align${Capitalize<Breakpoint>}`, typeof alignProp>;

const propsCol = breakpointProps(colProp, 'cols-{0}') as Record<`cols${Capitalize<Breakpoint>}`, typeof colProp>;

const propsGutters = breakpointProps(gutterProp, 'gutters-{0}') as Record<`gutters${Capitalize<Breakpoint>}`, typeof gutterProp>;

const propsGuttersX = breakpointProps(gutterProp, 'gutters-x-{0}') as Record<`guttersX${Capitalize<Breakpoint>}`, typeof gutterProp>;

const propsGuttersY = breakpointProps(gutterProp, 'gutters-y-{0}') as Record<`guttersY${Capitalize<Breakpoint>}`, typeof gutterProp>;

const propsJustify = breakpointProps(justifyProp, 'justify-{0}') as Record<`justify${Capitalize<Breakpoint>}`, typeof justifyProp>;
</script>

<script lang="ts" setup>
const props = defineProps({
    align: alignProp,
    ...propsAlign,
    cols: colProp,
    ...propsCol,
    gutters: gutterProp,
    ...propsGutters,
    guttersX: gutterProp,
    ...propsGuttersX,
    guttersY: gutterProp,
    ...propsGuttersY,
    justify: justifyProp,
    ...propsJustify,
    ...tagProps,
});

const alignProps = {
    sm: toRef(props, 'alignSm'),
    md: toRef(props, 'alignMd'),
    lg: toRef(props, 'alignLg'),
    xl: toRef(props, 'alignXl'),
    xxl: toRef(props, 'alignXxl'),
};

const alignClass = useStringTemplate('align-items-{0}', props.align).templatedString;

const alignClasses = useBreakpoints(alignProps, 'align-items-{0}-{1}').breakpointClasses;

const colProps = {
    sm: toRef(props, 'colsSm'),
    md: toRef(props, 'colsMd'),
    lg: toRef(props, 'colsLg'),
    xl: toRef(props, 'colsXl'),
    xxl: toRef(props, 'colsXxl'),
};

const colClass = useStringTemplate('row-cols-{0}', props.cols).templatedString;

const colClasses = useBreakpoints(colProps, 'row-cols-{0}-{1}').breakpointClasses;

const guttersProps = {
    sm: toRef(props, 'guttersSm'),
    md: toRef(props, 'guttersMd'),
    lg: toRef(props, 'guttersLg'),
    xl: toRef(props, 'guttersXl'),
    xxl: toRef(props, 'guttersXxl'),
};

const guttersClass = useStringTemplate('g-{0}', props.gutters).templatedString;

const guttersClasses = useBreakpoints(guttersProps, 'g-{0}-{1}').breakpointClasses;

const guttersXProps = {
    sm: toRef(props, 'guttersXSm'),
    md: toRef(props, 'guttersXMd'),
    lg: toRef(props, 'guttersXLg'),
    xl: toRef(props, 'guttersXXl'),
    xxl: toRef(props, 'guttersXXxl'),
};

const guttersXClass = useStringTemplate('gx-{0}', props.guttersX).templatedString;

const guttersXClasses = useBreakpoints(guttersXProps, 'gx-{0}-{1}').breakpointClasses;

const guttersYProps = {
    sm: toRef(props, 'guttersYSm'),
    md: toRef(props, 'guttersYMd'),
    lg: toRef(props, 'guttersYLg'),
    xl: toRef(props, 'guttersYXl'),
    xxl: toRef(props, 'guttersYXxl'),
};

const guttersYClass = useStringTemplate('gy-{0}', props.guttersY).templatedString;

const guttersYClasses = useBreakpoints(guttersYProps, 'gy-{0}-{1}').breakpointClasses;

const justifyProps = {
    sm: toRef(props, 'justifySm'),
    md: toRef(props, 'justifyMd'),
    lg: toRef(props, 'justifyLg'),
    xl: toRef(props, 'justifyXl'),
    xxl: toRef(props, 'justifyXxl'),
};

const justifyClass = useStringTemplate('justify-content-{0}', props.justify).templatedString;

const justifyClasses = useBreakpoints(justifyProps, 'justify-content-{0}-{1}').breakpointClasses;

const {classes} = useClasses(computed(() => [
    alignClass.value,
    ...alignClasses.value,
    colClass.value,
    ...colClasses.value,
    guttersClass.value,
    ...guttersClasses.value,
    guttersXClass.value,
    ...guttersXClasses.value,
    guttersYClass.value,
    ...guttersYClasses.value,
    justifyClass.value,
    ...justifyClasses.value,
]));
</script>
