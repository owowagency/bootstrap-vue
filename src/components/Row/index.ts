import Row from './Row.vue';
export default Row;
export * from './type';
