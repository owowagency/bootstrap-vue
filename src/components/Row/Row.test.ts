import Row from '.';

describe('template', () => {
    componentRenderTest(Row);

    componentSlotRenderTest(Row);

    componentWrapperTagTest(Row, {props: {tag: 'ul'}});

    componentWrapperClassTest(Row, {align: 'center'}, 'align-items-center');
    componentWrapperClassTest(Row, {alignSm: 'end'}, 'align-items-sm-end');
    componentWrapperClassTest(Row, {alignMd: 'start'}, 'align-items-md-start');
    componentWrapperClassTest(Row, {alignLg: 'center'}, 'align-items-lg-center');
    componentWrapperClassTest(Row, {alignXl: 'end'}, 'align-items-xl-end');
    componentWrapperClassTest(Row, {alignXxl: 'start'}, 'align-items-xxl-start');

    componentWrapperClassTest(Row, {cols: 1}, 'row-cols-1');
    componentWrapperClassTest(Row, {colsSm: 2}, 'row-cols-sm-2');
    componentWrapperClassTest(Row, {colsMd: 3}, 'row-cols-md-3');
    componentWrapperClassTest(Row, {colsLg: 4}, 'row-cols-lg-4');
    componentWrapperClassTest(Row, {colsXl: 5}, 'row-cols-xl-5');
    componentWrapperClassTest(Row, {colsXxl: 6}, 'row-cols-xxl-6');

    componentWrapperClassTest(Row, {gutters: 1}, 'g-1');
    componentWrapperClassTest(Row, {guttersSm: 2}, 'g-sm-2');
    componentWrapperClassTest(Row, {guttersMd: 3}, 'g-md-3');
    componentWrapperClassTest(Row, {guttersLg: 4}, 'g-lg-4');
    componentWrapperClassTest(Row, {guttersXl: 5}, 'g-xl-5');
    componentWrapperClassTest(Row, {guttersXxl: 6}, 'g-xxl-6');

    componentWrapperClassTest(Row, {guttersX: 1}, 'gx-1');
    componentWrapperClassTest(Row, {guttersXSm: 2}, 'gx-sm-2');
    componentWrapperClassTest(Row, {guttersXMd: 3}, 'gx-md-3');
    componentWrapperClassTest(Row, {guttersXLg: 4}, 'gx-lg-4');
    componentWrapperClassTest(Row, {guttersXXl: 5}, 'gx-xl-5');
    componentWrapperClassTest(Row, {guttersXXxl: 6}, 'gx-xxl-6');

    componentWrapperClassTest(Row, {guttersY: 1}, 'gy-1');
    componentWrapperClassTest(Row, {guttersYSm: 2}, 'gy-sm-2');
    componentWrapperClassTest(Row, {guttersYMd: 3}, 'gy-md-3');
    componentWrapperClassTest(Row, {guttersYLg: 4}, 'gy-lg-4');
    componentWrapperClassTest(Row, {guttersYXl: 5}, 'gy-xl-5');
    componentWrapperClassTest(Row, {guttersYXxl: 6}, 'gy-xxl-6');

    componentWrapperClassTest(Row, {justify: 'around'}, 'justify-content-around');
    componentWrapperClassTest(Row, {justifySm: 'between'}, 'justify-content-sm-between');
    componentWrapperClassTest(Row, {justifyMd: 'center'}, 'justify-content-md-center');
    componentWrapperClassTest(Row, {justifyLg: 'end'}, 'justify-content-lg-end');
    componentWrapperClassTest(Row, {justifyXl: 'evenly'}, 'justify-content-xl-evenly');
    componentWrapperClassTest(Row, {justifyXxl: 'start'}, 'justify-content-xxl-start');
});
