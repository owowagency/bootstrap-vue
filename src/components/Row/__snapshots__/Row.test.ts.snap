// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`template > renders default 1`] = `
"<div class=\\"row\\">
  <!-- @slot Displays content inside the row -->
</div>"
`;
